### ISSUE TITLE
sstable2json fails on sstable generated by scylla

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylla#1890]
Scylla version: c32803f2f08060edc921812543c80d099c80596e (master)

```
ERROR 14:15:54 Error in ThreadPoolExecutor
java.lang.ArithmeticException: / by zero
	at org.apache.cassandra.io.sstable.SSTable.estimateRowsFromIndex(SSTable.java:225) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.io.sstable.SSTableReader.buildSummary(SSTableReader.java:767) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.io.sstable.SSTableReader.load(SSTableReader.java:715) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.io.sstable.SSTableReader.load(SSTableReader.java:665) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.io.sstable.SSTableReader.open(SSTableReader.java:482) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.io.sstable.SSTableReader.open(SSTableReader.java:381) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.io.sstable.SSTableReader$4.run(SSTableReader.java:519) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[na:1.8.0_111]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_111]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[na:1.8.0_111]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_111]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_111]
ERROR 14:15:55 Error in ThreadPoolExecutor
java.lang.ArithmeticException: / by zero
	at org.apache.cassandra.io.sstable.SSTable.estimateRowsFromIndex(SSTable.java:225) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.io.sstable.SSTableReader.buildSummary(SSTableReader.java:767) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.io.sstable.SSTableReader.load(SSTableReader.java:715) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.io.sstable.SSTableReader.load(SSTableReader.java:665) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.io.sstable.SSTableReader.open(SSTableReader.java:482) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.io.sstable.SSTableReader.open(SSTableReader.java:381) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.io.sstable.SSTableReader$4.run(SSTableReader.java:519) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[na:1.8.0_111]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_111]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[na:1.8.0_111]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_111]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_111]
ERROR 14:15:55 Error in ThreadPoolExecutor
java.lang.ArithmeticException: / by zero
	at org.apache.cassandra.io.sstable.SSTable.estimateRowsFromIndex(SSTable.java:225) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.io.sstable.SSTableReader.buildSummary(SSTableReader.java:767) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.io.sstable.SSTableReader.load(SSTableReader.java:715) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.io.sstable.SSTableReader.load(SSTableReader.java:665) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.io.sstable.SSTableReader.open(SSTableReader.java:482) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.io.sstable.SSTableReader.open(SSTableReader.java:381) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.io.sstable.SSTableReader$4.run(SSTableReader.java:519) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[na:1.8.0_111]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_111]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[na:1.8.0_111]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_111]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_111]
ERROR 14:15:55 Corrupt sstable /var/lib/scylla/data/keyspace1/standard1-755cb640b10211e69420000000000000/keyspace1-standard1-ka-1=[Index.db, TOC.txt.tmp, Data.db]; skipped
java.io.EOFException: EOF after 8485 bytes out of 11690
	at org.apache.cassandra.io.util.FileUtils.skipBytesFully(FileUtils.java:391) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.db.RowIndexEntry$Serializer.skipPromotedIndex(RowIndexEntry.java:151) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.db.RowIndexEntry$Serializer.skip(RowIndexEntry.java:142) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.io.sstable.SSTable.estimateRowsFromIndex(SSTable.java:221) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.io.sstable.SSTableReader.buildSummary(SSTableReader.java:767) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.io.sstable.SSTableReader.load(SSTableReader.java:715) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.io.sstable.SSTableReader.load(SSTableReader.java:665) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.io.sstable.SSTableReader.open(SSTableReader.java:482) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.io.sstable.SSTableReader.open(SSTableReader.java:381) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at org.apache.cassandra.io.sstable.SSTableReader$4.run(SSTableReader.java:519) ~[apache-cassandra-2.1.8-SNAPSHOT.jar:2.1.8-SNAPSHOT]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_111]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_111]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_111]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_111]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_111]
ERROR 14:15:55 LEAK DETECTED: a reference (org.apache.cassandra.utils.concurrent.Ref$State@37e8814a) to class org.apache.cassandra.io.sstable.SSTableReader$InstanceTidier@959408583:/var/lib/scylla/data/system/peers-37f71aca7dc2383ba70672528af04d4f/system-peers-ka-4 was not released before the reference was garbage collected
ERROR 14:15:55 LEAK DETECTED: a reference (org.apache.cassandra.utils.concurrent.Ref$State@3d10ea3a) to class org.apache.cassandra.io.sstable.SSTableReader$InstanceTidier@2014471164:/var/lib/scylla/data/keyspace1/standard1-755cb640b10211e69420000000000000/keyspace1-standard1-ka-1 was not released before the reference was garbage collected
ERROR 14:15:55 LEAK DETECTED: a reference (org.apache.cassandra.utils.concurrent.Ref$State@b035f65) to class org.apache.cassandra.io.sstable.SSTableReader$InstanceTidier@1226866323:/var/lib/scylla/data/keyspace1/standard1-755cb640b10211e69420000000000000/keyspace1-standard1-ka-4 was not released before the reference was garbage collected
ERROR 14:15:55 LEAK DETECTED: a reference (org.apache.cassandra.utils.concurrent.Ref$State@73db1cb7) to class org.apache.cassandra.io.sstable.SSTableReader$InstanceTidier@406690995:/var/lib/scylla/data/keyspace1/standard1-755cb640b10211e69420000000000000/keyspace1-standard1-ka-3 was not released before the reference was garbage collected
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: 5001
	at org.apache.cassandra.io.sstable.metadata.MetadataSerializer.deserialize(MetadataSerializer.java:112)
	at org.apache.cassandra.io.sstable.metadata.MetadataSerializer.deserialize(MetadataSerializer.java:92)
	at org.apache.cassandra.io.sstable.SSTableReader.open(SSTableReader.java:455)
	at org.apache.cassandra.io.sstable.SSTableReader.open(SSTableReader.java:381)
	at org.apache.cassandra.io.sstable.SSTableReader.open(SSTableReader.java:376)
	at org.apache.cassandra.io.sstable.SSTableReader.open(SSTableReader.java:368)
	at org.apache.cassandra.tools.SSTableExport.export(SSTableExport.java:369)
	at org.apache.cassandra.tools.SSTableExport.export(SSTableExport.java:382)
	at org.apache.cassandra.tools.SSTableExport.main(SSTableExport.java:466)

```

**Comment by slivne** [comment]
@tgrabiec can you either:
1.  provide the schema of the sstable and the inserts that cause a creation of a "bad sstable"
or 
2. provide the raw files if possible.

I am not sure its clear to you - but is this a regression ?

**Comment by tgrabiec** [comment]
@slivne It's default cassandra-stress write workload. Probably a regression, because I recall that sstable2json used to work, but I haven't verified right now.

**Comment by tgrabiec** [comment]
The number 5001 in one of the errors matches our `metadata_type::Sharding`. Could be caused by 919825a2c7eacf5d56b108a4e29010e72f3d1ee9

\cc @avikivity

**Comment by slivne** [comment]
I am bisecting I think its the additional data @avikivity added to the
summary files

will have an answer in ~10 minutes

**Comment by slivne** [comment]
confirmed

the issues is with

commit 919825a2c7eacf5d56b108a4e29010e72f3d1ee9
Merge: 18b9fa3 07d5a20
Author: Paweł Dziepak <pdziepak@scylladb.com>
Date:   Wed Nov 23 11:25:53 2016 +0000

    Merge "Improve sharding in large clusters" from Avi

    "Clusters with a large number of nodes, or a low number of vnodes, and a
    high number of shards, or a combination, suffer from an aliasing
problem:
    both vnodes and intra-node sharding consider the most significant bits
    to select the owning node and owning shard respectively.  Since the same
    bits are used for both, a low number of vnodes leads to some shards
    being overcommitted relative to others.

    This series fixes the problem by sharding on bits 0:47 of the token
    (murmur3 partitioner only), leaving the most significant 12 bits for
    vnodes.  Simulation shows that this value provides reasonable sharding
    for 100-node, 30-shard clusters.

    In order to prevent re-sharding sstables on each boot, token ranges for
    the range are stored in a new sub-component of the sstable Statistics
    component. With the default 12 ignored bits we have 4096 token ranges
    for non-Level-compacted SSTables, which takes some space but is still
    reasonable.

    Fixes #1277."


the commit before that works

**Comment by avikivity** [comment]
It's trivial to fix, but our sstables won't be readable by Cassandra.  I guess we'll have to switch to using a new component, which is sad, I wanted to avoid adding more files.

In `org.apache.cassandra.io.sstable.metadata.MetadataSerializer`,

```java
        Map<MetadataType, Integer> toc = new HashMap<>(numComponents);
        for (int i = 0; i < numComponents; i++)
        {
            toc.put(MetadataType.values()[in.readInt()], in.readInt());
        }
```

Since the new component is not a member of the MetadataType enum, Java complains.

**Comment by slivne** [comment]
fixed by 0bc3ce7e09dc99d1f1a39223408fb1a6d8a0aa33
